### ISSUE TITLE
Give `StreamBuilder` ability to be lossless

### ISSUE THREAD

**Issue by gaaclarke** [flutter/flutter#112197]
The documentation for `StreamBuilder` clearly states that the `builder` isn't called 1-to-1 with the `Stream` and that its intention is to just display the latest snapshot.  However that doesn't stop people from assuming that it is 1-to-1 as evidenced in the StackOverflow answer here: https://stackoverflow.com/questions/64167602/create-list-view-with-stream-builder?newreg=44e616311fef471588b8f25184be3a6a.

I suspect the name of the class and its signature communicates that it may behave different than it does.  A more proper name may have been `LatestSnapshotBuilder`.  The punishment for using the API incorrectly is too high in this case.  Imagine shipping a product that behaves correctly on your internet then a user that has a fast internet connection uses it and data is missing from the list.

Since the fail case is so dire I think we should have an escape valve in the API design, a `bool` that guarantees that `StreamBuilder` calls the builder 1-to-1 with the stream events.  We could rename the class but that would probably be too disruptive.

Proposed API:
```dart
  const StreamBuilder({
    super.key,
    this.initialData,
    super.stream,
    required this.builder,
    bool isLossless = false,  // <-- new parameter
  });
```

**Comment by bernaferrari** [comment]
You should check the "swallow error" property from StreamBuilder/FutureBuilder, maybe isLossless could be made on the same way, a global variable, so users can override everywhere instead individually.

**Comment by sarannng** [comment]
Hi, @gaaclarke, I would like to work on this fix. If you can assign this to me it would be great. Thanks.

**Comment by gaaclarke** [comment]
Hey @sarannng, sounds good.  Feel free to ping me for a PR review.  It should be a fun fix.  I'm not going to assign it to you since the team typically doesn't do that for PRs like this.  You are free to fix the issue though.

**Comment by sarannng** [comment]
Hi @gaaclarke, just to clarify, all I need to do is add isLossLess parameter to StreamBuilder, right?

**Comment by gaaclarke** [comment]
That's correct, and make it work with tests showing it works of course.

**Comment by goderbauer** [comment]
How is the `isLossless` param supposed to work? If the stream provides updates more frequently then the framerate you still wouldn't see each entry on screen.

**Comment by gaaclarke** [comment]
> How is the `isLossless` param supposed to work? If the stream provides updates more frequently then the framerate you still wouldn't see each entry on screen.

Right, but we'd play all the updates we've received from the Stream, even if that meant wasted building of things that will just be replaced before they are viewed.  The implementation now is throwing them away.  The `isLossless` argument would make the sample code from the stackoverflow link in the description work.

The important thing is, does the Stream represent atomic state, or does it represent partial state.  Right now StreamBuilder works with atomic state Streams, but doesn't for partial state.

Even if no one used `isLossless` it would be worth implementing it imo since it would be another opportunity to communicate the semantics of StreamBuilder, and the only communication that happens in code.

**Comment by sultanhunter** [comment]
@gaaclarke is anybody working on this? Can I pickup this as my first contribution?

**Comment by gaaclarke** [comment]
There was someone interested in the past but I didn't see a PR.  No one on the team is working on it so it's available.  If you get a PR you can ping me to take a first look.

**Comment by sultanhunter** [comment]
@gaaclarke in the given stackoverflow description why the error goes away on wrapping the Streambuilder with Scaffold widget and how it is related to the streambuilder's builder not gettting called 1-to-1 with the Stream?

**Comment by gaaclarke** [comment]
It's because of the mutation of the `items` variable.  See how items are getting added to it in the `else` block?  Because the StreamBuilder callback isn't guaranteed to be executed for each item, there is a race condition here with the contents of `items` can be different depending on how fast the Stream executes.

**Comment by sultanhunter** [comment]
@gaaclarke yeah I understood that builder callback is not getting called 1-to-1 with the stream events. But the given stackoverflow question will still throw same error (No Material widget found) if we use Listview.builder in place of StreamBuilder.

**Comment by sultanhunter** [comment]
@gaaclarke by directly calling the build method instead of using setState I am able to run the stackoverflow code without any errors with Streambuilder's builder callback getting called properly 1-to-1 with the stream events with proper Connection states, but the UI gets stuck at circularprogress indicator as framework does not actually rebuilds the UI if we just call the build method directly. Can you guide me how to solve this problem. I read the setState code and it marks and schedule the widget for rebuild, but the framework does not provide any function which we can call directly in our widget to rebuild the UI.

<img width="1071" alt="Screenshot 2022-10-25 at 8 43 17 PM" src="https://user-images.githubusercontent.com/39861073/197813079-fb1f02b2-27d8-4cc9-86de-fcc337d9d349.png">

**Comment by goderbauer** [comment]
> Maybe you should consider capturing the data then calling the builder closure for each piece of data and throwing away the result until you get to the last value. You can try doing that in the lambda getting passed to setState.

This sounds ... wasteful and confusing to build UI that just gets thrown away.

I would instead suggest either providing all items that have accumulated since the last build to the builder so it can decide what it wants to do with them. Or, they could be provided to an accumulation function that can process them before the single result is passed to the builder. Not sure how to integrate this into the current API though.

**Comment by gaaclarke** [comment]
> This sounds ... wasteful and confusing to build UI that just gets thrown away.

Wasteful yea, I wouldn't say it's confusing though.  What is demonstrably confusing (by the stackoverflow example) is not calling the closure for every piece of data.

> Or, they could be provided to an accumulation function that can process them

That sounds good to me too.  Sounds like a breaking change though...  Unless you created a different builder argument that has a different signature.  That would eliminate the constructor argument for lossy or not, the new closure represents if you want it called losslessly.  We could potentially deprecate the old field too if we wanted.

```dart
class StreamBuilder<T> {
  StreamBuilder(this.builder, this.losslessBuilder);
  final Widget Function(T state) builder;
  final Widget Function(List<T> states) losslessBuilder;
}
```

**Comment by sultanhunter** [comment]
> It's because of the mutation of the `items` variable. See how items are getting added to it in the `else` block? Because the StreamBuilder callback isn't guaranteed to be executed for each item, there is a race condition here with the contents of `items` can be different depending on how fast the Stream executes.

The reason which I understood for not guaranteeing Streambuilder's builder callback getting executed for each stream events:
because of setState scheduling builds for two stream events which happens very fast resulting in only calling build method for the latest event which in turn only executes the builder callback for the latest event.
<img width="1382" alt="Screenshot 2022-10-26 at 12 26 11 AM" src="https://user-images.githubusercontent.com/39861073/197858526-89c2cb49-2f5c-4ea5-ac48-32c1b7953170.png">

Let me know if this is correct.

**Comment by gaaclarke** [comment]
> because of setState scheduling builds for two stream events which happens very fast resulting in only calling build method for the latest event which in turn only executes the builder callback for the latest event.

That's correct.  Another way to think about it is that if the stream's frequency is greater than half the build frequency, data will be lost (because of the nyquist frequency).  Also, the build frequency isn't fixed so the slower the builds, the more likely we'll lose data.

**Comment by sultanhunter** [comment]
@gaaclarke this is the solution which I tried to apply, but wasn't able to solve the problem 🙃:
1-  created a list (streamData) which stores all the incoming events data from the stream. 
2-  we will call the setState for the first element of the streamData list.
3- when the build method will get executed, we will remove the first element from the streamData list, then we will check if the streamData list is still not empty. If it is not empty we will call the setState again for the first element of the streamData list.
4- on stream done callback, we will check if the streamData list is still not empty. If its not empty we will recursively call the setState for all the items in the streamData list.

![Screenshot 2022-10-26 at 2 07 41 AM](https://user-images.githubusercontent.com/39861073/197879148-27cfc0c2-b86b-45b6-9bd4-64e0d62fefa9.jpg)

![Screenshot 2022-10-26 at 2 07 08 AM](https://user-images.githubusercontent.com/39861073/197879119-0322a44a-39c0-443b-8e14-b5e5be28b4a8.jpg)

Please let me know what I am doing wrong.

**Comment by gaaclarke** [comment]
One thing that looks fishy is you are adding to the end of streamData but pulling from the front.  You are also only removing one item per build.  You'd want to clear out everything by the end of the build.  I haven't played around with the code to know exactly how to best make that happen.  If I did I'd have implemented the solution myself =)

**Comment by sultanhunter** [comment]
@gaaclarke one more question:
Since the build rate is slower than the stream events, and we are caching the stream data and building for each data, that means stream events will not be sync with build, as a result it will take more time to build the latest snapshot widget as comparison to current lossy streambuilder implementation. Is that correct?

**Comment by gaaclarke** [comment]
> @gaaclarke one more question: Since the build rate is slower than the stream events, and we are caching the stream data and building for each data, that means stream events will not be sync with build, as a result it will take more time to build the latest snapshot widget as comparison to current lossy streambuilder implementation. Is that correct?

Yea, technically true but in practice it will be negligible.  If we do the suggestion about about having a lossless callback the only extra overhead will be adding to List the values.  Building will just look at the latest one if that's what it wants.

**Comment by sultanhunter** [comment]
@gaaclarke but why when build rates are faster then streams and builder callback is getting called 1-to-1 with stream events in current streamBuilder implementation but we still getting same error in the stackoverflow code?

**Comment by PrahladPawar1733** [comment]
Hey, I am Prahlad and learning about contributing to an open source, but I am new to this, so can anyone help me?

**Comment by CaleyD** [comment]
The problem in the referenced stackoverflow is a fundamental misunderstanding of Flutter and the real solution would be removing state mutations from the build method. The proposal in this issue might solve a fraction of the probelem but then imagine what happens when the developer updates the code to look up a value on an InheritedWodget in that builder function and is then frustrated their builder is getting called too often and there are excess values in the ‘items’ array. 

Hacking in the framework doesn’t seem like a healthy solution to this particular education problem.

**Comment by caseycrogers** [comment]
Forcing build to be called on obsolete values doesn't sound very coherent.
If you need accumulated values, wrap your stream in an accumulator. I wrote this package awhile back for this:
https://pub.dev/packages/stream_summary_builder

I'm not sure, but I think `RxDart` also has equivalents.

I agree that the current API is really confusing and misleading, but I think we'll have a better time communicating to users the limitations of the API than overcoming them in the widget.

That said, I guess you could add an optional "fold" argument to `StreamBuilder`:
```dart
/// A function that folds update to stream builder into a single value.
`T Function(List<T>, T) fold;`
```
Since `fold` is independent of the build cycle, it can be happily called on every event coming off the stream. It would be a little kludgy to use for accumulating a list:
```dart
// This is off the top of my head, may have trivial syntax errors.
return StreamBuilder<List<Foo>>(
  // Have to coerce the stream events to a list.
  // We could avoid this by adding a second generic to `StreamBuilder`
  // so that fold and build can use a different type than the stream,
  // but that comes with its own downsides.
  stream: myDb.getFoos().map((foo) => [foo]),
  builder: (context, fooLstSnap) {
    return ListView(
      children: fooLstSnap.requireData.map(
        (foo) => FooView(foo),
      ).toList(),
    );
  }
);

**Comment by prabinlamsal19** [comment]
@gaaclarke  Can I pickup this issue as my first contribution?

**Comment by YuriyRum** [comment]
Hello everyone,

I would also like to contribute an idea. What if we add a new parameter "duration" to the StreamBuilder interface. And in the _subscribe method instead of changing state in the stream listener, we can collect data into the queue and then every interval (which is defined by the duration parameter) pass every piece of data to the state. In this case, the app developer decides how often the state is changed. 
Below is a code sample, it is not a working code, it just expresses an idea

 ```
   void _subscribe() {
    if (widget.stream != null) {
      _subscription = widget.stream!.listen((T data) {
        queue.add(data);
      }, onError: (Object error, StackTrace stackTrace) {
        setState(() {
          _summary = widget.afterError(_summary, error, stackTrace);
        });
      }, onDone: () {
        setState(() {
          _summary = widget.afterDone(_summary);
        });
      });
      _summary = widget.afterConnected(_summary);
    }
    Timer.periodic(widget.duration, (_) {
      if (mounted) {
        if(queue.isNotEmpty) {
          setState(() {
            _summary = queue.removeAt(0);
          });
        }
      }
    });
  }
```

**Comment by RotRot-pi** [comment]
@gaaclarke Hello,

I noticed that the StackOverflow issue you referenced seems to come from a misunderstanding of Flutter’s widget hierarchy rather than a limitation of the `StreamBuilder` itself. 
Given that, I’m curious to know how you got the idea of proposing the 'isLossless`?  
is this common enough issue to justify modifying the API ?

**Comment by gaaclarke** [comment]
> Given that, I’m curious to know how you got the idea of proposing the `isLossless`?
> is this common enough issue to justify modifying the API ?

1) There are things a developer can know by reading an API and there are things a developer can know by reading the docstrings.  Information is more efficiently conveyed through the API since everyone has to read that in order for it to be used.
1) The default implicit contract is "data will not be lost", whenever that is not the case it should be loudly advertised.
1) This is especially insidious because you can run it and get correct results sometimes.

**Comment by caseycrogers** [comment]
I agree the API as it exists today is a massive foot cannon-and a very common one. This specific beginner trap wasted hours of my time when I was first learning Flutter.

However, the proposed solution is not very coherent as the conversation thus far has shown-a core tenant of Flutter that enables the framework to work effectively is that (with some very specific exceptions like lazy-loading) business logic should not be triggered from the build cycle. Putting business logic into `StreamBuilder` violates this very core tenant. When user expectations don't match the reality of the API, it's great to just bring the API in-line with user expectations, but here that just plainly doesn't make sense.

My earlier proposal (https://github.com/flutter/flutter/issues/112197#issuecomment-1538983941) lays out a way to use the API to guide beginners away from this footgun without putting logic into the build function. However, even then I don't think it's a particularly good approach as `reduce` better belongs as an extension method on Stream (as I believe RXDart and some other packages provide) rather than as special case logic embedded in `StreamBuilder`.

Generally, I think the real answer is that `StreamBuilder` and `FutureBuilder` are both fundamentally foot guns (IIRC Hixie and others have said they have regrets about creating these widgets) and most senior Flutter devs I know just "unwrap" futures and streams in their state management layer (see [ex](https://github.com/alesalv/flutter_arch_comp/blob/43c8d89d4ebcfd314269d8e8af11745d5203a39b/lib/src/pokemon/controllers/pokemon_controller.dart#L15)). Namely, neither `Future` nor `Stream` give you synchronous access to the current value. This creates problems like losing the current value for a stream or getting a first frame loading flash when initializing a `FutureBuilder` with an already completed Future.

As the example above shows, unwrapping a `Future`/`Stream` is a bit tedious. I think the best approach is for Flutter to provide core primitives that make it easier to do so and then guide developers towards these approaches and away from the Future/Stream builders. Here's my issue proposing exactly this:
https://github.com/flutter/flutter/issues/135047

In the meantime, I'd highly recommend removing the `good first issue` tag on this issue as it's clearly not a good first issue and a lot of beginners looking for their first contribution are getting side tracked trying to implement a solution here.
